Demo 2 Quarto notebook tutorial for PEcAn uncertainty analysis (#3570)
PecanProject/pecan: documentation/tutorials/Demo_02_Uncertainty_Analysis/uncertainty.qmd

3. You can run the code chunks sequentially
4. Once you have successfully run the demo, you can modify parameters to configure new runs and analyses

**Objective:**

This demo illustrates how to run an uncertainty analysis workflow in PEcAn using an R-based Quarto notebook. The uncertainty workflow includes ensemble analysis and sensitivity analysis to comprehensively assess model uncertainty and parameter importance. It will cover loading settings, writing model configuration files, running model simulations, performing ensemble-based uncertainty quantification, and conducting sensitivity analysis to identify which parameters most influence model results. This approach provides a programmatic alternative to the web-based PEcAn interface for executing ecosystem models with advanced uncertainty analysis capabilities.

# 2. PEcAn Version

This section displays the version information for all PEcAn packages.

```{r version-info}
PEcAn.all::pecan_version()
```

# 3. Install SIPNET v1.3.0

If you haven't already installed the SIPNET binary, you can do so by running the following code. This will download the SIPNET binary to `demo_outdir/sipnet` and make it executable.

> > Note: The `demo_outdir` directory will be created in the root of your PEcAn installation (i.e., at `documentation/tutorials/Demo_2_Uncertainty_Analysis/demo_outdir/`). This directory will contain the SIPNET binary as well as the output generated by PEcAn in this demo.

Reviewer: - `documentation/tutorials/Demo_2_Uncertainty_Analysis/demo_outdir/` is not at root of PEcAn installation.
- do we need separate demo_outdir for the two demos? 
- can we use `withr::local_tempdir()`?

Author: Basically `documentation/tutorials/Demo_1_Basic_Run/demo_outdir` is the output for demo 1 notebook and 
`documentation/tutorials/Demo_2_Uncertainty_Analysis/demo_outdir/` is the output directory for demo 2 uncertainty notebook.

> do we need separate demo_outdir for the two demos?

I believe it’s better to have separate directories for each notebook so that users can easily identify which notebook generated which outputs.

> can we use withr::local_tempdir()?

We could do that, but in this case it would be harder for new users to locate the generated outputs. Using a constant directory name makes it easier for users to know exactly where the outputs are being stored.

```{r run-sensitivity-analysis}
# Execute ensemble and sensitivity analyses to quantify uncertainty and parameter influence
# First, retrieve results from all model runs
runModule.get.results(settings)

# Run ensemble analysis: generates ensemble predictions and uncertainty quantification
runModule.run.ensemble.analysis(settings)

# Run sensitivity analysis: generates variance decomposition plots and sensitivity metrics
runModule.run.sensitivity.analysis(settings)
```

# 12. Understanding PEcAn Uncertainty Analysis Outputs

After running ensemble and sensitivity analyses, PEcAn generates several important outputs that help you understand model uncertainty and parameter sensitivity.

## 12.1 Ensemble Analysis Outputs

The ensemble analysis produces:

- **`ensemble.Rdata`**: Contains `ensemble.output` object with model predictions for all ensemble members
- **`ensemble.analysis.[RunID].[Variable].[StartYear].[EndYear].pdf`**: Histogram and boxplot of ensemble predictions
- **`ensemble.ts.[RunID].[Variable].[StartYear].[EndYear].pdf`**: Time-series plot showing ensemble mean, median, and 95% confidence intervals

> ## 12.2 Sensitivity Analysis Outputs

Reviewer: add basic SA, VD, ENS plots ... https://github.com/PecanProject/pecan/blob/develop/modules/uncertainty/R/plots.R
<img width="418" height="227" alt="image" src="https://github.com/user-attachments/assets/401f39ba-e74e-4275-9d47-e0452ae23aa2" />
<img width="544" height="312" alt="image" src="https://github.com/user-attachments/assets/23c23965-e6f8-47c5-ad6e-6eb81a985bed" />

Author: Added Plots in section 11 in the notebook.

# 13. Customizing Ensemble Analysis Parameters (Optional)

## (Optional) Use this section only if you want to override the default ensemble analysis parameters. Skip if defaults are sufficient.
> **Important:** If you modify the ensemble analysis parameters in this section, you must re-run sections 8, 9, and 10 to regenerate the model output with the new ensemble configuration.

```{r customize-ensemble}
# Set the number of ensemble members (model runs)
settings$ensemble$size <- 50

# Specify the variable to be analyzed in the ensemble
settings$ensemble$variable <- "NEE"

# Choose the method for sampling the parameter space (options: "uniform", "lhc", "halton", "sobol", "torus")
settings$ensemble$samplingspace$parameters$method <- "halton"
```

# 14. Customizing Sensitivity Analysis Parameters (Optional)

## (Optional) Use this section only if you want to override the default sensitivity analysis parameters. Skip if defaults are sufficient.
> **Important:** If you modify the sensitivity analysis parameters in this section, you must re-run sections 8, 9, and 10 to regenerate the model output with the new sensitivity analysis configuration.

```{r customize-sensitivity}

> # Set the quantiles (in standard deviations) for parameter distribution in sensitivity analysis

Reviewer: define what -2:2 mean @dlebauer

```{r explore-settings}
str(settings)
```

Your turn: explore the `settings` object further using the following commands:
*   `names(settings)` to list the top-level keys in the settings object
*   Once you know the names, you can look at each component in detail, for example:
*   `settings$run` to access the run-specific settings, such as start and end dates, model type, and output directory
*   `settings$pfts` to explore the Plant Functional Types settings

Now you can update each of these settings. Here is a simple example:

```{r add-info-to-settings}
settings$info <- list(
  author = "Aritra Dey",
  date = Sys.Date(),
  description = "Demo run of PEcAn using SIPNET"
)
```

Editing the more interesting settings to change the PFT (`settings$pfts`) or extend the run (`settings$run$end.date`) is an optional exercise for the user. For example, you could change the pft or the end date, but this requires additional parameter files and meteorological inputs. This includes files containing parameters for that PFT (`settings$pfts$pft$posterior.files`), or a climate file (`settings$run$met$path$path1`) that extends to the desired simulation period.

> # 7. Create Output Directory

Reviewer: @AritraDey-Dev I think that this is automated by the `prepare.settings()` function, so the section can be removed. Please confirm.

Author: yes indeed  this is not needed.

**What this notebook does:**

1. Configure a PEcAn workflow by loading and validating a `pecan.xml` settings file.
2. Run a set of ecosystem model simulations by writing model configuration files and then running SIPNET.
3. Quantify uncertainty using ensemble analysis, sensitivity analyses, and variance decomposition.
4. Visualize results to identify important parameters and how they influence model variance.
5. Change configuration settings and re-run the workflow.

## Prerequisites

To run this notebook, you will need to install PEcAn and its dependencies, as well as download the SIPNET model binary.

### PEcAn packages and dependencies.

```
# Enable repository from pecanproject
options(repos = c(
  pecanproject = 'https://pecanproject.r-universe.dev',
  CRAN = 'https://cloud.r-project.org'))
install.packages('PEcAn.all')
```

A valid `pecan.xml` configuration file. Start with the example at `pecan/documentation/tutorials/Demo_02_Uncertainty_Analysis/pecan.xml`.

### SIPNET v1.3.0

If you haven't already installed the SIPNET model, you can do so by running the following code. This will download the SIPNET binary to `demo_outdir/sipnet` and make it executable.

> > Note: The `demo_outdir` directory will be created in the root of your PEcAn installation (i.e., at `documentation/tutorials/Demo_02_Uncertainty_Analysis/demo_outdir/`). This directory will contain the SIPNET binary as well as the output generated by PEcAn in this demo.

Reviewer 2: ```suggestion
> Note: The `demo_outdir` directory will be created inside of your PEcAn installation, at `documentation/tutorials/Demo_02_Uncertainty_Analysis/demo_outdir/`. This directory will contain the SIPNET binary as well as the output generated by PEcAn in this demo.
```

(For the record I don't love this outdir placement, but I think we already dicussed that in the PR for demo 1. My chnge here is just to describe the existing placement more clearly)

## Settings

Example settings for this demo live at `pecan/documentation/tutorials/Demo_02_Uncertainty_Analysis/pecan.xml` and you can read more about the settings in the [PEcAn Documentation](https://pecanproject.github.io/pecan-documentation/develop/pecanXML.html#pecanXML), and sections focused on [ensemble](https://pecanproject.github.io/pecan-documentation/develop/xml-advanced.html#xml-ensemble) and [sensitivity analysis](https://pecanproject.github.io/pecan-documentation/develop/xml-advanced.html#xml-sensitivity-analysis) settings in particular.

Open that settings file and look at the ensemble and sensitivity analysis sections. You can modify these settings to change the number of ensemble members, the variable to analyze, and the sampling method. Some of the key settings in this demo include:

**Ensemble size**

The number of runs in the ensemble is set to **50** in this demo, but a larger ensemble size (100-5000) is often used in practice to better estimate the posterior distribution of the output.

**Output variable**

> The output variable for the ensemble analysis is set to **NPP** (Net Primary Productivity). You can change this to other variables like NEE (Net Ecosystem Exchange), LAI (Leaf Area Index), ET (Evapotranspiration), etc.

Reviewer 2: Would it be more useful or more distracting to mention here that you can set multiple outputs?

Author: I believe that would definitely be useful. @infotroph, please confirm whether I should add a comment about multiple outputs.

Author: Added in https://github.com/PecanProject/pecan/pull/3570/commits/628f626f6bf4d09c65939c6c620d2873f13c7e77

**Sampling method**

The sampling method for generating parameter sets is set to **halton** in this demo. This tells PEcAn to sample using the [Halton sequence](https://en.wikipedia.org/wiki/Halton_sequence), which is a quasi-random sampling method that more efficiently explores parameter space than random sampling. Other options include Latin Hypercube (lhc), which is another quasi-random sequence, as well as uniform that draws random samples. A random sampler requires a larger ensemble size to adequately explore parameter space.

**Sensitivity analysis quantiles**

> PEcAn's sensitivity analysis includes a handy shortcut that converts a specified standard deviation into its normal quantile equivalent. In the example pecan.xml, these are set to **set -1,1** (the median value, 0, occurs by default) which are converted internally to the 15.9th and 84.1th quantiles of the parameter distribution. You can add more quantiles (e.g., -2, 2) to explore a wider range of parameter values, and -3:3 are often used in practice. 

Reviewer 2: ```suggestion
PEcAn's sensitivity analysis includes a handy shortcut that converts a specified standard deviation into its normal quantile equivalent. In the example pecan.xml, these are set to **-1, 1** (the median value, 0, occurs by default) which are converted internally to the 15.9th and 84.1th quantiles of the parameter distribution. You can add more quantiles to explore a wider range of parameter values: {-3, -2, -1, 1, 2, 3} is often used in practice. 
```

> By working in quantiles relative to each parameter’s distribution, the sensitivity and variance decomposition reflect sensitivity across the range of parameter values. Many sensitivity analyses use a fixed perturbation size such as the mean +/- 10%. This approach does not capture the uncertainty across the parameter distribution and can not be used for variance decomposition.

Reviewer 2: Is "Many sensitivity analyses use a fixed perturbation" talking about what's commonly done for analyses done inside PEcAn, or what's commonly done for non-PEcAn SAs but discouraged by this tutorial? If the latter, does this clarify?

```suggestion
By working in quantiles relative to each parameter’s distribution, the sensitivity and variance decomposition reflect sensitivity across the range of parameter values. Many sensitivity analyses tools use a fixed perturbation size such as the mean +/- 10%. PEcAn's SA does not take this approach because it does not capture the uncertainty across the parameter distribution and can not be used for variance decomposition.
```

Author: I intended this as a general comment on sensitivity analysis.

## Load the settings file

```{r load-settings}
settings_path <- here::here("documentation/tutorials/Demo_02_Uncertainty_Analysis/pecan.xml")

settings <- PEcAn.settings::read.settings(settings_path)

settings <- PEcAn.settings::prepare.settings(settings)
```

See Demo 1 Section 6 for details on what these functions do. Briefly, they read the XML file, convert it into an R list object that PEcAn can use, check that settings are valid, fill in defaults, and create the output directory.

# 6. Write Model Configuration Files

This step generates the model-specific configuration files that will be used to run the ecosystem model. The process involves:

1. Disabling database write operations because we are not using a database
    ```{r disable-db-write}

>    settings$database <- NULL # Disable database operations for this demo

Reviewer 2: As we [discussed for demo 1](https://github.com/PecanProject/pecan/pull/3531#pullrequestreview-3085926968), I don't see why this is needed -- setting a list entry to NULL is the same as removing it, so why not just delete it from the XML in the first place?) -- and I'm really quite sure the user _should_ not need to do it. If deleting this line makes the demo work wrong, please file an issue with a reprex so that we can fix whatever makes this distracting hack necessary.

Author: Yes, for now deleting this line will not make the demo work. I agree that this line is unrelated to the demo. This is probably an issue with the PEcAn DB package.

```{r run-ensemble-analysis}
runModule.run.ensemble.analysis(settings)
```

**Sensitivity Analysis**: Systematically varies individual parameters to assess their influence on model outputs. This identifies which parameters most strongly affect model predictions and helps prioritize parameter refinement efforts.

```{r run-sensitivity-analysis}
runModule.run.sensitivity.analysis(settings)
```

# 9. Output 

## Output organization

These are the key folders and files that will be created under the directory defined by `settings$outdir` (e.g., `demo_outdir` in the example). The file contents are described in the next section.

We have discussed the structure of the output directory in Demo 1 (Section 10), but now we have three new folders that contain outputs from the sensitivity, ensemble, and variance decomposition analyses. Here we focus on the additional outputs generated by the ensemble and sensitivity analyses.

```
demo_outdir/
├── run/                   # Configuration & execution metadata
│   ├── runs.txt           # List of run IDs (one per model realization)

> │   ├── <runid>/           # Model-specific config copies (sometimes)

Reviewer 2: ```suggestion
│   ├── <runid>/           # Files for each individual model invocation, in the form the model saw them
```

> │   └── config.*           # Generated model configs (e.g., SIPNET)

Reviewer 2: I have never seen `config.*` in the rundir! What module generates them?

Author: Yes,A lot of refactoring need to be done in this part.i will fix this.

├── out/                   # Raw model outputs by run ID

> │   └── <runid>/           # E.g., daily or sub-daily SIPNET output files

Reviewer 2: ```suggestion
├── out/                   # Model outputs by run ID
│   └── <runid>/           # E.g., daily or sub-daily SIPNET output files + yearly netcdfs + log files from the run
```

├── sensitivity.analysis/  # Sensitivity raw outputs &amp; figures
│   ├── sensitivity.output.*.Rdata
│   └── sensitivity.analysis.*.pdf
├── ensemble.analysis/     # Ensemble summaries &amp; figures
│   ├── ensemble.Rdata
│   ├── ensemble.analysis.*.pdf
│   └── ensemble.ts.*.pdf
├── variance.decomposition/  # Created when variance decomposition runs

> │   └── variance.decomposition.*.pdf

Reviewer 2: I'm used to seeing all of these saved directly in outdir without their own subdirectories -- what process creates these?

Author: fixed in https://github.com/PecanProject/pecan/pull/3570/commits/97c15b9ad11327ccf65d7654f0959206ceb2553c

  main = paste("Carbon Pools Over Time — PEcAn", runid)
)
lines(model_output$posix, model_output$TotSoilCarb, col = "brown")
legend("topright", legend = c("Total Live Biomass", "Total Soil Carbon"), col = c("darkgreen", "brown"), lty = 1)
```

## 17.3 Plot Water Variables

```{r plot-water-variables}
# Plot Soil Moisture and Snow Water Equivalent
plot(model_output$posix, model_output$SoilMoist,
  type = "l",
  col = "blue",
  xlab = "Date",
  ylab = "Soil Moisture (kg m-2)",
  main = paste("Soil Moisture Over Time — PEcAn", runid)
)
lines(model_output$posix, model_output$SWE, col = "lightblue")
legend("topright", legend = c("Soil Moisture", "Snow Water Equivalent"), col = c("blue", "lightblue"), lty = 1)
```

## 17.4 Plot LAI and Biomass

```{r plot-lai-biomass}
# Plot Leaf Area Index (LAI) and Above Ground Wood
plot(model_output$posix, model_output$LAI,
  type = "l",
  col = "darkgreen",
  xlab = "Date",
  ylab = "LAI (m2 m-2)",
  main = paste("Leaf Area Index Over Time — PEcAn", runid)
)
lines(model_output$posix, model_output$AbvGrndWood, col = "brown")
legend("topright", legend = c("LAI", "Above Ground Wood"), col = c("darkgreen", "brown"), lty = 1)
```

> # 18. PEcAn Outputs

Author: This part is almost same as section 9 which tells about pecan outputs,also in this part as it mentions trait and meta analysis which we are not doing in this notebook,so can we remove this part @dlebauer ?

Reviewer: I will remove this part and push the changes. 

I've created https://github.com/PecanProject/pecan/issues/3661 to track future re-implementation of a meta-analysis tutorial.

Reviewer: And have implemented the requested revisions in https://github.com/AritraDey-Dev/pecan/pull/10
